### ISSUE TITLE
refactor import results in corrupted source code if the first line is an import statement

### ISSUE THREAD

**Issue by Hiroyuki-Nagata** [ensime/ensime-server#1029]
I executed a useful command `C-x C-r t` , but it sometimes fail its import.
In my case, I tried to import like below:
```
import org.scalatra._
import javax.servlet.ServletContext
import _root_.akka.actor.ActorSystem
import com.mordor.minas_morgul.Foo
import com.mordor.minas_morgul.Bar
import com.mordor.minas_morgul.Yahoo
```

...and, exec `C-x C-r t`, but it failed. `import` is bizarre.
```
iimport com.mordor.minas_morgul.Google
mport org.scalatra._
import javax.servlet.ServletContext
import _root_.akka.actor.ActorSystem
import com.mordor.minas_morgul.Foo
import com.mordor.minas_morgul.Bar
import com.mordor.minas_morgul.Yahoo
```

Logs are like below:
```
16:04:18.408 [default-akka.actor.default-dispatcher-38] INFO  o.e.s.protocol.swank.SwankProtocol - Received msg: (:swank-rpc (swank:typecheck-file (:file "/home/golumn/git_lotr_tools/lotr-strategy/src/main/scala/ScalatraBootstrap.scala" :contents "iimport com.mordor.servlet.Swagg\
 erServlet\nmport org.scalatra._\nimport javax.servlet.ServletContext\nimport _root_.akka.actor.ActorSystem\nimport com.mordor.dao.ConnectionInfoFactory\nimport com.mordor.statsd.StatsdClient\nimport com.mordor.lotr.strategy._\nimport com.mordor.lotr.strategy.service.{StrategyDisp\
 atchClient, Strateg...
 16:04:18.412 [default-akka.actor.default-dispatcher-48] INFO  o.e.s.protocol.swank.SwankProtocol - Writing: (:clear-all-scala-notes)
 16:04:18.416 [Scala Presentation Compiler []] WARN  org.ensime.util.PresentationReporter - Unsupported operation during reporting
 java.lang.UnsupportedOperationException: Position.start on class scala.reflect.internal.util.OffsetPosition
         at scala.reflect.internal.util.Position.start(Position.scala:114) ~[scala-reflect-2.10.2.jar:na]
         at org.ensime.util.PresentationReporter.info0(Reporter.scala:42) ~[ensime_2.10.jar:0.9.10-SNAPSHOT]
         at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply$mcV$sp(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.reporters.Reporter.withoutTruncating(Reporter.scala:42) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.reporters.Reporter.error(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.CompilationUnits$CompilationUnit.error(CompilationUnits.scala:102) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.ast.parser.Parsers$UnitParser$$anonfun$showSyntaxErrors$2.apply(Parsers.scala:226) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.ast.parser.Parsers$UnitParser$$anonfun$showSyntaxErrors$2.apply(Parsers.scala:225) [scala-compiler-2.10.2.jar:na]
         at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772) [scala-library-2.10.2.jar:na]
         at scala.collection.immutable.List.foreach(List.scala:318) [scala-library-2.10.2.jar:na]
         at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32) [scala-library-2.10.2.jar:na]
         at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45) [scala-library-2.10.2.jar:na]
         at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771) [scala-library-2.10.2.jar:na]
         at scala.tools.nsc.ast.parser.Parsers$UnitParser.showSyntaxErrors(Parsers.scala:225) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.ast.parser.Parsers$UnitParser.smartParse(Parsers.scala:242) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.ast.parser.SyntaxAnalyzer$ParserPhase.apply(SyntaxAnalyzer.scala:29) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:464) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.Global$Run$$anonfun$compileLate$2$$anonfun$apply$1.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.Global$Run$$anonfun$compileLate$2$$anonfun$apply$1.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
         at scala.reflect.internal.SymbolTable.atPhase(SymbolTable.scala:207) [scala-reflect-2.10.2.jar:na]
         at scala.tools.nsc.Global$Run$$anonfun$compileLate$2.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.Global$Run$$anonfun$compileLate$2.apply(Global.scala:1682) [scala-compiler-2.10.2.jar:na]
         at scala.collection.Iterator$class.foreach(Iterator.scala:727) [scala-library-2.10.2.jar:na]
         at scala.collection.AbstractIterator.foreach(Iterator.scala:1157) [scala-library-2.10.2.jar:na]
         at scala.tools.nsc.Global$Run.compileLate(Global.scala:1682) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.interactive.Global.scala$tools$nsc$interactive$Global$$parseAndEnter(Global.scala:573) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.interactive.Global.backgroundCompile(Global.scala:480) [scala-compiler-2.10.2.jar:na]
         at scala.tools.nsc.interactive.PresentationCompilerThread.run(PresentationCompilerThread.scala:25) [scala-compiler-2.10.2.jar:na]
 16:04:18.417 [Scala Presentation Compiler []] WARN  org.ensime.util.PresentationReporter - Unsupported operation during reporting
16:04:18.417 [Scala Presentation Compiler []] WARN  org.ensime.util.PresentationReporter - Unsupported operation during reporting
java.lang.UnsupportedOperationException: Position.start on class scala.reflect.internal.util.OffsetPosition
        at scala.reflect.internal.util.Position.start(Position.scala:114) ~[scala-reflect-2.10.2.jar:na]
        at org.ensime.util.PresentationReporter.info0(Reporter.scala:42) ~[ensime_2.10.jar:0.9.10-SNAPSHOT]
        at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply$mcV$sp(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.reporters.Reporter$$anonfun$error$1.apply(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.reporters.Reporter.withoutTruncating(Reporter.scala:42) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.reporters.Reporter.error(Reporter.scala:70) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.CompilationUnits$CompilationUnit.error(CompilationUnits.scala:102) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.ast.parser.Parsers$UnitParser$$anonfun$showSyntaxErrors$2.apply(Parsers.scala:226) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.ast.parser.Parsers$UnitParser$$anonfun$showSyntaxErrors$2.apply(Parsers.scala:225) [scala-compiler-2.10.2.jar:na]
        at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772) [scala-library-2.10.2.jar:na]
        at scala.collection.immutable.List.foreach(List.scala:318) [scala-library-2.10.2.jar:na]
        at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32) [scala-library-2.10.2.jar:na]
        at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45) [scala-library-2.10.2.jar:na]
        at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771) [scala-library-2.10.2.jar:na]
        at scala.tools.nsc.ast.parser.Parsers$UnitParser.showSyntaxErrors(Parsers.scala:225) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.ast.parser.Parsers$UnitParser.smartParse(Parsers.scala:242) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.ast.parser.SyntaxAnalyzer$ParserPhase.apply(SyntaxAnalyzer.scala:29) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:464) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.Global$Run$$anonfun$compileLate$2$$anonfun$apply$1.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.Global$Run$$anonfun$compileLate$2$$anonfun$apply$1.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
        at scala.reflect.internal.SymbolTable.atPhase(SymbolTable.scala:207) [scala-reflect-2.10.2.jar:na]
        at scala.tools.nsc.Global$Run$$anonfun$compileLate$2.apply(Global.scala:1683) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.Global$Run$$anonfun$compileLate$2.apply(Global.scala:1682) [scala-compiler-2.10.2.jar:na]
        at scala.collection.Iterator$class.foreach(Iterator.scala:727) [scala-library-2.10.2.jar:na]
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1157) [scala-library-2.10.2.jar:na]
        at scala.tools.nsc.Global$Run.compileLate(Global.scala:1682) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.interactive.Global.scala$tools$nsc$interactive$Global$$parseAndEnter(Global.scala:573) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.interactive.Global.backgroundCompile(Global.scala:480) [scala-compiler-2.10.2.jar:na]
        at scala.tools.nsc.interactive.PresentationCompilerThread.run(PresentationCompilerThread.scala:25) [scala-compiler-2.10.2.jar:na]
```

**Comment by fommil** [comment]
These logs are after the importing happens. Can you give a fuller example of where the point is when you ask to import. Also update Scala to the latest stable in that series if possible.

**Comment by Vifon** [comment]
Some more information is available in this issue: https://github.com/ensime/ensime-server/issues/1029

An **incomplete** unittest is available on my fork, on branch `import-test`: https://github.com/Vifon/ensime-server/tree/import-test. I don't know how to make the call to the refactoring functions so that the imports will be sorted like in the call from Emacs.

From what I've gathered, the issue only occurs when the new import is added on the very first line of the file (or on the first byte even), so the sorting is necessary for it to occur.

**Comment by fommil** [comment]
renamed to be less emacs-specific, this sounds like a server bug by Vifon's recent analysis.

**Comment by Hiroyuki-Nagata** [comment]
I checked the @Vifon 's repository.
There is a problem, if ensime-server processed import some packages and it would be inserted the first line, `AddImportRefactorDesc` would work. But, `AddImportRefactorDesc` don't process like `OrganiseImportsRefactorDesc`, in other words, the import order will be odd. If the function working like both functions exist ( do import and sort ), it will be solved.

[Refactoring.scala#L239](https://github.com/ensime/ensime-server/blob/master/core/src/main/scala/org/ensime/core/Refactoring.scala#L239)
```
        case OrganiseImportsRefactorDesc(file) =>
          reloadAndType(file)
          doOrganizeImports(procId, tpe, file)
        case AddImportRefactorDesc(qualifiedName, file) =>
          reloadAndType(file)
          doAddImport(procId, tpe, qualifiedName, file)
```

I tried to make `doAddImport` working as also `doOrganizeImports`, but that is very complicated...
